File: LeetCodeProblemsCMake/Problems/LeetCode3637TrionicArrayI.cpp
//
// Created by Saeed Jannati on 2/3/2026 A.
//

#include "LeetCode3637TrionicArrayI.h"
using namespace std;

bool LeetCode3637TrionicArrayI::isTrionic(vector<int> &nums) {
    int previous{nums[0]};
    vector<int> slopes{};
    int slope{};
    int delta{};
    for (int i = 1, e = nums.size(); i < e; i++) {
        delta = nums[i] - previous;
        previous = nums[i];
        if (delta == 0) {
            return false;
        }
        if (slopes.size() == 0) {
            if (delta < 0) {
                return false;
            }
            slopes.push_back(1);
            continue;
        }
        slope = delta > 0 ? 1 : -1;
        if (slope * slopes[slopes.size() - 1] > 0) {
            continue;
        }
        slopes.push_back(slope);
        if (slopes.size() > 3) {
            return false;
        }
    }
    return slopes.size() == 3;
}

File: LeetCodeProblemsGo/LeetCode3637TrionicArrayI/LeetCode3637TrionicArrayI.go
package LeetCode3637TrionicArrayI

func isTrionic(nums []int) bool {
	slopes := []int{}
	previous := nums[0]
	length := len(nums)
	slope := 0
	delta := 0
	for i := 1; i < length; i++ {
		delta = nums[i] - previous
		previous = nums[i]
		if delta == 0 {
			return false
		}
		if len(slopes) == 0 {
			if delta < 0 {
				return false
			}
			slopes = append(slopes, 1)
			continue
		}
		if delta > 0 {
			slope = 1
		} else {
			slope = -1
		}
		if slope*slopes[len(slopes)-1] > 0 {
			continue
		}
		slopes = append(slopes, slope)
		if len(slopes) > 3 {
			return false
		}
	}
	if len(slopes) != 3 {
		return false
	}
	return true
}
